File: Regression/Random_forest_regression/random_forest_regression.R
# Random Forest Regression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Random Forest Regression to the Training Set
library(randomForest)
set.seed(1234)
# dataset[n] will give a data frame, dataset$var will give a vector
regressor = randomForest(x = dataset[1],
                         y = dataset$Salary,
                         ntree = 500
                         )

# Predicting a new result with Random Forest Regression
y_pred = predict(regressor, data.frame(Level = 6.5))


# Visualizing the Random Forest results (higher res and smoother curves)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level = x_grid))),
            color = 'blue') +
  ggtitle('Truth or Bluff (Decision Tree Regression)') +
  xlab('Level') +
  ylab('Salary')

File: Regression/Random_forest_regression/random_forest_regression.py
# Random Forest Regression
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into training set and test set
#from sklearn.model_selection import train_test_split
## test_size is a small percentage = 20% - 40%
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2,
#                                                    random_state = 0)

# Feature Scaling
#from sklearn.preprocessing import StandardScaler
#sc_X = StandardScaler()
#X_train = sc_X.fit_transform(X_train)
#X_test = sc_X.transform(X_test)

# Fitting the Random Forest Regression Model to the dataset
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(n_estimators = 300, random_state = 0)
regressor.fit(X, y)

# Predicting a new result with Random Forest Regression
y_pred = regressor.predict([[6.5]])

# Visualizing the Random Forest Regression results (for higher res and smoother curve)
plt.figure(1)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title('Truth or Bluff (Random Forest Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()


plt.figure(2)
x_ln = np.arange(0, 1, 0.0001)
y_ln = np.log(x_ln / (1 - x_ln))
plt.plot(x_ln, y_ln, color = 'blue')
plt.xlim(-1, 5)
plt.ylim(-10, 10)
plt.show()
